### PR TITLE
MDEV-15951 system versioning by trx id doesn't work with partitioning

### PR DESCRIPTION
**Pull request by FooBarrior**
MDEV-15951 system versioning by trx id doesn't work with partitioning

Fix partitioning for trx_id-versioned tables.
`partition by hash`, `range` and others now work.
`partition by system_time` is forbidden.
Currently we cannot use row_start and row_end in `partition by`, because
insertion of versioned field is done by engine's handler, as well as 
row_start/row_end's value set up, which is a transaction id -- so it's 
also forbidden.

The drawback is that it's now impossible to use `partition by key()`
without parameters for such tables, because it references row_start and
row_end implicitly.

* add handler::vers_can_native()
* drop Table_scope_and_contents_source_st::vers_native()
* drop partition_element::find_engine_flag as unused
* forbid versioning partitioning for trx_id as not supported
* adopt vers tests for trx_id partitioning
* forbid any row_end referencing in `partition by` clauses,
  including implicit `by key()`

### CODE REVIEW

File: sql/ha_partition.h [hunk 1]
```diff
@@ -413,6 +413,16 @@ class ha_partition :public handler
 
   virtual void return_record_by_parent();
 
+  bool vers_can_native(THD *thd) override
```
Comment (kevgs) [inline sql/ha_partition.h]: C++11 is not allowed, IIRC.

File: sql/ha_partition.h [hunk 2]
```diff
@@ -413,6 +413,16 @@ class ha_partition :public handler
...
+  {
+    // PARTITION BY SYSTEM_TIME is not supported for now
+    bool can= !thd->lex->part_info
+              || thd->lex->part_info->part_type != VERSIONING_PARTITION;
+    for (int i= 0; i < m_tot_parts && can; i++)
```
Comment (kevgs) [inline sql/ha_partition.h]:
```
In file included from /work/mariadb/sql/sql_partition_admin.cc:32:
/work/mariadb/sql/ha_partition.h:421:22: warning: comparison of integers of different signs: 'int' and 'uint' (aka 'unsigned int') [-Wsign-compare]
    for (int i= 0; i < m_tot_parts && can; i++)
                   ~ ^ ~~~~~~~~~~~
```

**Review by kevgs — COMMENTED** [review]
```sql
versioning.alter 'innodb'                w1 [ fail ]
        Test ended at 2018-08-02 15:37:41

CURRENT_TEST: versioning.alter
--- /home/kevg/work/mariadb/mysql-test/suite/versioning/r/alter.result	2018-08-02 15:29:32.666228578 +0300
+++ /home/kevg/work/mariadb/mysql-test/suite/versioning/r/alter.reject	2018-08-02 15:37:41.017261932 +0300
@@ -41,7 +41,7 @@
 add column trx_end bigint(20) unsigned as row end invisible,
 add period for system_time(trx_start, trx_end),
 add system versioning;
-ERROR HY000: `trx_start` must be of type TIMESTAMP(6) for system-versioned table `t`
+ERROR HY000: `trx_start` must be of type TIMESTAMP(6) for system-versioned table `(temporary)`
 alter table t
 add column trx_start timestamp as row start invisible,
 add column trx_end timestamp as row end invisible,
```

**Comment by kevgs** [comment]
```sql
--source include/have_innodb.inc
--source include/have_partition.inc

create table t (
  a int primary key,
  row_start bigint unsigned as row start invisible,
  row_end bigint unsigned as row end invisible,
  period for system_time(row_start, row_end)
) engine=innodb with system versioning
# partition by key() (
#   partition p1,
#   partition p2
# )
;

ins--source include/have_innodb.inc
--source include/have_partition.inc

create table t (
  a int primary key,
  row_start bigint unsigned as row start invisible,
  row_end bigint unsigned as row end invisible,
  period for system_time(row_start, row_end)
) engine=innodb with system versioning
# partition by key() (
#   partition p1,
#   partition p2
# )
;

insert into t values (1);

delete from t where a=1;
set @re=100500;
select row_end from t for system_time all where a=1 into @re;

select *,row_end from t for system_time all where a=1 and row_end=@re;

drop table t;
ert into t values (1);

delete from t where a=1;
set @re=100500;
select row_end from t for system_time all where a=1 into @re;

select *,row_end from t for system_time all where a=1 and row_end=@re;

drop table t;

```

When I'm uncommenting partition definition the last `SELECT` returns nothing. That happens because it looks for a row in a wrong partition.

Partition by `KEY` use `PRIMARY KEY` which consists of (a, row_end). When It inserts a row it determines a partition number in `get_part_id_key()`. And it's `MAX_BIGINT` on `INSERT`. When we remove a row we insert historical one. It's `row_end` is not max value. It's 37 or something in my case. But partition engine do not know it because 37 is set inside InnoDB. Eventually it put historical row in a wrong partition.

When I set `partition by key(a)` all works fine.

To sum up, `ha_partition` uses incorrect `row_end` value before InnoDB set the correct one.

**Review by kevgs — CHANGES_REQUESTED** [review]
Anything should be done with that `partition by key`. Warnings and failing test should be fixed too.

**Comment by midenok** [comment]
Please, fix requested changes and rebase on latest trunk.

**Comment by FooBarrior** [comment]
Thanks for the report, @kevgs.
There's no interface in `handler` for fixig this bug smoothly, and i feel it's quite connected with _partitioning by `trx_id` + vers_, which is a separate feature. So i am going just to forbid using `row_end` in `partition by` statement clauses.

**Review by midenok — APPROVED** [review]
Please fix Eugene's comments and merge.

**Comment by kevgs** [comment]
Sorry I haven't thought about that earlier but `row_start` seems to be affected too for the same reason: on `INSERT` its value is not initialized by InnoDB.

```sql
--source include/have_innodb.inc
--source include/have_partition.inc

create or replace table t (
  a int,
  row_start bigint unsigned as row start invisible,
  row_end bigint unsigned as row end invisible,
  period for system_time(row_start, row_end)
) engine=innodb with system versioning
partition by hash(row_start) (
  partition p1,
  partition p2
);

insert into t values (1);
insert into t values (2);

set @rs=100500;
select row_start from t for system_time all where a=1 into @rs;
delete from t;

select *, row_start from t for system_time all where a=1 and row_start=@rs;

drop table t;

```

I have a crash in this test. If you haven't try adding more rows - this changes InnoDB transaction id.

```
Thread 1 (Thread 0x7fa6f5525700 (LWP 10038)):
#0  __pthread_kill (threadid=<optimized out>, signo=6) at pthread_kill.c:57
#1  my_write_core (sig=6) at stacktrace.c:481
#2  handle_fatal_signal (sig=6) at signal_handler.cc:305
#3  <signal handler called>
#4  __GI_raise (sig=sig@entry=6) at raise.c:51
#5  __GI_abort () at abort.c:79
#6  __assert_fail_base (fmt=0x7fa6fbddb7d8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x1699b45 "part_id == m_last_part", file=file@entry=0x1698776 "/work/mariadb/sql/ha_partition.cc", line=line@entry=4522, function=function@entry=0x1699bf1 "virtual int ha_partition::delete_row(const uchar *)") at assert.c:92
#7  __GI___assert_fail (assertion=0x1699b45 "part_id == m_last_part", file=0x1698776 "/work/mariadb/sql/ha_partition.cc", line=4522, function=0x1699bf1 "virtual int ha_partition::delete_row(const uchar *)") at assert.c:101
#8  ha_partition::delete_row (this=0x7fa6a0121b50, buf=0x7fa6a008e3c8 "\375\002") at ha_partition.cc:4522
#9  handler::ha_delete_row (this=0x7fa6a0121b50, buf=0x7fa6a008e3c8 "\375\002") at handler.cc:6292
#10 mysql_delete (thd=0x7fa6a0000cf8, table_list=0x7fa6a0013eb8, conds=0x0, order_list=0x7fa6a0005548, limit=18446744073709551614, options=0, result=0x0) at sql_delete.cc:768
#11 mysql_execute_command (thd=0x7fa6a0000cf8) at sql_parse.cc:4917
#12 mysql_parse (thd=0x7fa6a0000cf8, rawbuf=0x7fa6a0013df0 "delete from t", length=13, parser_state=0x7fa6f5524560, is_com_multi=false, is_next_command=false) at sql_parse.cc:8077

```

**Comment by FooBarrior** [comment]
Truly so :disappointed:
So row_end is now also forbidden.
I've also removed the new error code in favor of already existing (although, surprisingly, unused) `ER_VERS_NOT_SUPPORTED`

**Review by kevgs — APPROVED** [review]
LGTM
